test/integration/scheduler/plugins/plugins_test.go

```diff
@@ -2561,7 +2442,34 @@ func initTestSchedulerForFrameworkTest(t *testing.T, testCtx *testutils.TestCont
 			t.Fatal(err)
 		}
 	}
-	return testCtx
+
+	// teardown will delete created resources and scheduler here, resources should be those
+	// that will affect the scheduling result, like nodes, pods, etc.. Namespaces should not be
+	// deleted here because it's created together with the apiserver, they should be deleted
+	// simultaneously or we'll have no namespace.
+	// This should only be called when you want to kill the scheduler alone, away from apiserver.
+	// For example, in scheduler integration tests, recreating apiserver is performance consuming,
+	// then shutdown the scheduler and recreate it between each test case is a better approach.
+	teardown := func() {
+		err := testCtx.ClientSet.CoreV1().Nodes().DeleteCollection(testCtx.SchedulerCtx, *metav1.NewDeleteOptions(0), metav1.ListOptions{})
+		if err != nil {
+			t.Errorf("error while deleting all nodes: %v", err)
+		}
+		err = testCtx.ClientSet.CoreV1().Pods(testCtx.NS.Name).DeleteCollection(testCtx.SchedulerCtx, *metav1.NewDeleteOptions(0), metav1.ListOptions{})
+		if err != nil {
+			t.Errorf("error while deleting pod: %v", err)
+		}
+		// Wait for all pods to be deleted, or will failed to create same name pods
+		// required in other test cases.
+		if err := wait.Poll(time.Millisecond, wait.ForeverTestTimeout,
+			testutils.PodsCleanedUp(testCtx.SchedulerCtx, testCtx.ClientSet, testCtx.NS.Name)); err != nil {
+			t.Errorf("error while waiting for all pods to be deleted: %v", err)
```

Comment:
This will result on the infamous "timed out waiting for the condition" error when something goes wrong.

When waiting for some condition, Gomega is better. Its failure message will include information about the latest error (if there was any while checking the condition) and the last negative condition check result.

https://github.com/kubernetes/kubernetes/pull/116207/commits/fc8198c08ffce91aa95f2a2fb30d7d352feac79f shows how to use `gomega.NewGomegaWithT(b).Eventually(...).Should(gomega.BeEmpty())` to check that pods, namespaces and claims got removed.

Integration tests are very similar to E2E tests in this regard, the only difference is that they are written with `go test` instead of Ginkgo.

I wonder whether it would be possible to consolidate... but probably not now. :sweat_smile:

Reply:
Let me track this with another issue, I think this is not an isolated case, we'd better to cover them once and for all.
And it's a big one already...